commit b0ba45dc55dfe0f803a050e37d4b885810a904a8
Author: Miroslav Rezanina <mrezanin@redhat.com>
Date:   Wed Jul 10 14:21:53 2013 +0200

    qga: escape cmdline args when registering win32 service (CVE-2013-2231)
    
    RH-Author: Laszlo Ersek <lersek@redhat.com>
    Message-id: <1372954619-30116-6-git-send-email-lersek@redhat.com>
    O-Subject: [EMBGD RHEL-6.4.z qemu-kvm PATCH 5/5] qga: escape cmdline args when registering win32 service (CVE-2013-2231)
    Bugzilla: 980758
    RH-Acked-by: Eric Blake <eblake@redhat.com>
    RH-Acked-by: Petr Matousek <pmatouse@redhat.com>
    RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>
    
    Reported-by: Lev Veyde <lveyde@redhat.com>
    Reviewed-by: Eric Blake <eblake@redhat.com>
    Acked-by: Michael Roth <mdroth@linux.vnet.ibm.com>
    Signed-off-by: Laszlo Ersek <lersek@redhat.com>
    
    Conflicts (when cherry picking from RHEL-6.5):
    
            qga/service-win32.c
    
    Because RHEL-6.4.z lacks RHEL-6.5's
    
      commit d013faa7efad1340df61aa0645b5b6d2dcde2b52
      Author: Laszlo Ersek <lersek@redhat.com>
      Date:   Fri Jun 7 12:23:58 2013 +0200
    
          qga: save state directory in ga_install_service()
    
    which is also upstream a839ee77c786a8200c76ca92f697eebf6bcc9aa3.

diff --git a/qga/service-win32.c b/qga/service-win32.c
index 94089a1..7c29a77 100644
--- a/qga/service-win32.c
+++ b/qga/service-win32.c
@@ -35,12 +35,75 @@ static int printf_win_error(const char *text)
     return n;
 }
 
+/* Windows command line escaping. Based on
+ * <http://blogs.msdn.com/b/oldnewthing/archive/2010/09/17/10063629.aspx> and
+ * <http://msdn.microsoft.com/en-us/library/windows/desktop/17w5ykft%28v=vs.85%29.aspx>.
+ *
+ * The caller is responsible for initializing @buffer; prior contents are lost.
+ */
+static const char *win_escape_arg(const char *to_escape, GString *buffer)
+{
+    size_t backslash_count;
+    const char *c;
+
+    /* open with a double quote */
+    g_string_assign(buffer, "\"");
+
+    backslash_count = 0;
+    for (c = to_escape; *c != '\0'; ++c) {
+        switch (*c) {
+        case '\\':
+            /* The meaning depends on the first non-backslash character coming
+             * up.
+             */
+            ++backslash_count;
+            break;
+
+        case '"':
+            /* We must escape each pending backslash, then escape the double
+             * quote. This creates a case of "odd number of backslashes [...]
+             * followed by a double quotation mark".
+             */
+            while (backslash_count) {
+                --backslash_count;
+                g_string_append(buffer, "\\\\");
+            }
+            g_string_append(buffer, "\\\"");
+            break;
+
+        default:
+            /* Any pending backslashes are without special meaning, flush them.
+             * "Backslashes are interpreted literally, unless they immediately
+             * precede a double quotation mark."
+             */
+            while (backslash_count) {
+                --backslash_count;
+                g_string_append_c(buffer, '\\');
+            }
+            g_string_append_c(buffer, *c);
+        }
+    }
+
+    /* We're about to close with a double quote in string delimiter role.
+     * Double all pending backslashes, creating a case of "even number of
+     * backslashes [...] followed by a double quotation mark".
+     */
+    while (backslash_count) {
+        --backslash_count;
+        g_string_append(buffer, "\\\\");
+    }
+    g_string_append_c(buffer, '"');
+
+    return buffer->str;
+}
+
 int ga_install_service(const char *path, const char *logfile)
 {
     int ret = EXIT_FAILURE;
     SC_HANDLE manager;
     SC_HANDLE service;
     TCHAR module_fname[MAX_PATH];
+    GString *esc;
     GString *cmdline;
     SERVICE_DESCRIPTION desc = { (char *)QGA_SERVICE_DESCRIPTION };
 
@@ -49,14 +112,18 @@ int ga_install_service(const char *path, const char *logfile)
         return EXIT_FAILURE;
     }
 
-    cmdline = g_string_new(module_fname);
-    g_string_append(cmdline, " -d");
+    esc     = g_string_new("");
+    cmdline = g_string_new("");
+
+    g_string_append_printf(cmdline, "%s -d",
+                           win_escape_arg(module_fname, esc));
 
     if (path) {
-        g_string_append_printf(cmdline, " -p %s", path);
+        g_string_append_printf(cmdline, " -p %s", win_escape_arg(path, esc));
     }
     if (logfile) {
-        g_string_append_printf(cmdline, " -l %s -v", logfile);
+        g_string_append_printf(cmdline, " -l %s -v",
+                               win_escape_arg(logfile, esc));
     }
 
     g_debug("service's cmdline: %s", cmdline->str);
@@ -85,6 +152,7 @@ out_manager:
 
 out_strings:
     g_string_free(cmdline, TRUE);
+    g_string_free(esc, TRUE);
     return ret;
 }
 
